Implemented API HTTP requests for Game list (GET/POST) and detail view (GET/PATCH/DELETE)

abrahmasandra opened the pull request:
Added an "api" app to chigame and added urls and their corresponding views using the Django REST framework. Specifically implemented the list view for Game objects, as well as the detail view under the endpoints `/api/games` and `/api/games/<int:pk>` respectively.

ENDPOINTS:

`/api/games/`: This supports both GET and POST requests. GET will return a list of all the game objects in the database using Django REST Framework. POST allows users to create new games

`/api/games/<int:gameID>`: This supports GET/PATCH/DELETE requests through Django REST framework. GET returns a detail view for a specific game, which PATCH allows a user to edit the contents of the game, and DELETE allows user to delete a certain game.

TASK COMPLETED:

When the endpoints are finished being implemented -- meaning all the functionality described above is complete.

abrahmasandra committed:
created API app

abrahmasandra committed:
changed api app location #48

abrahmasandra committed:
changed api location #48

abrahmasandra committed:
added rest_framework to installed apps #48

abrahmasandra committed:
created list and detail views (and corresponding url endpoints) for games #48 #50

abrahmasandra committed:
tweaked urls.py file of the main django project

abrahmasandra committed:
tweaked urls.py file of the main django project

fixed the merge from dev

abrahmasandra committed:
tweaked urls.py file of the main django project

fixed the merge from dev

abrahmasandra committed:
tweaked urls.py file of the main django project

fixed merge i hope

abrahmasandra committed:
created tests for API list and detail views using Django REST framework APITestCase packages #122

cuyakonwu reviewed (approved):
Code is working at http://127.0.0.1:8000/apis/games/, GET, POST, PUT, DELETE work as intended

cuyakonwu committed:
Merge branch 'dev' into apis/game-endpoints

cuyakonwu committed:
Updated djangorestframework #50

cuyakonwu committed:
Merge branch 'apis/game-endpoints' of github.com:uchicago-cs/chigame into apis/game-endpoints

majorsylvie commented on src/chigame/api/serializers.py:
```diff
@@ -0,0 +1,9 @@
+class GameSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Game
+        fields = ("name", "description", "min_players", "max_players")
```
As of now, neither the list nor detail views include the necesary information to uniquely identify a given game.

You don't include the primary key, nor an API url for games.
This means that once a user has listed games, they have no way to actually access the individual detail view or send actions like DELETE or PATCH.

Add in at least the primary key as a field such that the necessary identificatory information is always associated with each game.

You can think of this as, API users don't have access to a nice watchbuddies style game listing where you can click on links to do things like edit and delete. 
Thus you need to give the endpoint user all the info they need to reconstruct that kind of usually internally handled work that GUI and backend can do.

abrahmasandra replied:
I added the `id` field to the GameSerializer so that the user will also receive the primary key of games, when getting the list of games at the endpoint (/api/games). This will allow the user to use the `id` to then get the detail view of a specific game.

majorsylvie commented on src/chigame/api/tests.py:
```diff
@@ -0,0 +1,117 @@
+        self.assertEqual(Game.objects.get().name, "The Witcher 3: Wild Hunt")
+        self.assertEqual(Game.objects.get().description, "Geralt of Rivia")
+        self.assertEqual(Game.objects.get().min_players, 1)
+        self.assertEqual(Game.objects.get().max_players, 5)
```
Given that you're doing these kind of checks a lot, it might be useful to consider a helper function which can test some given fields to see if they have some associated values. 

That way you could pass in a more compact structure like a dictionary of expected fields and values, then that dictionary could be fed to your helper function which generically checks the provided fields via `Game.objects.get().[field_name]`.

majorsylvie commented on src/chigame/api/urls.py:
```diff
@@ -0,0 +1,8 @@
+
+urlpatterns = [
+    path("games/", views.GameListView.as_view(), name="game-list"),
+    path("games/<int:pk>/", views.GameDetailView.as_view(), name="game-detail"),
```
Nice job actually providing names. 
Consider adding `api-` in front of your names here, since these names are used outside the scope of just your app, so this could get confused with a `games` app URL.

Instead something like `api-game-list` is unmistakably the API's url.

abrahmasandra replied:
added the "api-" prefix before the names of the API views

majorsylvie commented on src/chigame/api/views.py:
```diff
@@ -0,0 +1,20 @@
+
+# Create your views here.
+
+# create /games/ endpoint that returns a list of all games
```
these comments can be removed.

abrahmasandra replied:
removed the comments in views.py

majorsylvie commented on src/chigame/users/migrations/0002_group_userprofile_groupinvitation_friendinvitation.py:
```diff
@@ -0,0 +1,97 @@
+# Generated by Django 4.2.4 on 2023-10-30 16:54
```
Both of your migration files are duplicates of files already on dev and cause issues with django migrations when merged with dev. 

Remove them from this PR and conflicts will go away. 

If you go to the github actions corresponding to this PR, you can see the tests fail, they fail because of there being multiple versions of the same migration. Which it asks you do fix with `python manage.py makemigrations --merge`
which will fail unless you actually delete these files / never include them in this commit meant for dev.

abrahmasandra replied:
I just deleted migration files

majorsylvie reviewed (changes requested):
Overall, the endpoints are close to being ready to go! 
Include the primary key and/or api url per-game-model in list and detail view.
Remove the migration files
And potentially refactor some of the test code.

Overall, this is going to set a strong precedent for the team and you're close to ready to merge!

abrahmasandra committed:
added the primary y_key to game list view #50

abrahmasandra committed:
delete comments in views.py and changed names urls

abrahmasandra committed:
deleted migration files #50

abrahmasandra committed:
deleted migration files #50

abrahmasandra committed:
updated test urls #50

abrahmasandra committed:
added helper function to check equality #50

abrahmasandra committed:
Merge branch 'dev' into apis/game-endpoints

abrahmasandra committed:
delete migrations

abrahmasandra committed:
fixed url file

abrahmasandra committed:
Delete src/chigame/games/migrations/0002_lobby.py

abrahmasandra committed:
Delete src/chigame/users/migrations/0001_initial.py

abrahmasandra committed:
Delete src/chigame/users/migrations/0003_added_notification_model.py

abrahmasandra committed:
Delete src/chigame/users/migrations/0004_added_visible_column_to_notifications.py

abrahmasandra committed:
Delete src/chigame/users/migrations/0005_added_and_changed_names_of_cols_in_notifications.py

abrahmasandra committed:
Delete src/chigame/users/migrations/0006_add_type_attribute_to_Notification.py

abrahmasandra committed:
migrations

abrahmasandra committed:
na

abrahmasandra commented:
I have fixed all the commented issues. Now the code passes the CI tests as well

majorsylvie commented on src/chigame/api/tests.py:
```diff
@@ -0,0 +1,136 @@
+        """
+
+        for key in expected:
+            self.assertEqual(getattr(object, key), expected[key])
```
Nice job implementing the testing helper function!

majorsylvie commented on src/chigame/api/tests.py:
```diff
@@ -0,0 +1,136 @@
+
+        # Create a game object
+        url = reverse("api-game-list")
+        data = {"name": "The Witcher 3", "description": "Geralt of Rivia", "min_players": 1, "max_players": 4}
```
Small note, for the sake of readability you might want to put each KV pair on a new line:

In general this is more readable and easily extensible since you can just add a new line, instead of having an ever expanding single line. 

This will matter much more as the `Game` model gets updated with new attributes being added

```
data = {
  "name": "The Witcher 3", 
  "description": "Geralt of Rivia", 
  "min_players": 1, 
  "max_players": 4
  }
```

majorsylvie commented on src/chigame/api/tests.py:
```diff
@@ -0,0 +1,136 @@
+
+        # Create a game object
+        url = reverse("api-game-list")
+        data = {"name": "The Witcher 3", "description": "Geralt of Rivia", "min_players": 1, "max_players": 4}
```
Given that you're reusing this same dictionary multiple times, you may want to create a helper factory like in the Django testing notes: 

https://canvas.uchicago.edu/courses/51666/files/10210480?wrap=1&fd_cookie_set=1

This will be especially relevant as the scale of the game related endpoints grows and you'll want more robust and easier to write tests.

majorsylvie reviewed (approved):
Great job taking all the feedback previously mentioned!

Adding the primary keys on the endpoint is very good.
I am very proud of the amount of tests already made for this endpoint!!

I have some smaller things of note for the tests of this endpoint, though these could be turned into a separate issue and PR of adding more tests and adding factories to create a game / games.

Thank you for your work and keep moving forward!

abrahmasandra commented:
Closing statement: created the API app within chigame, so that users can run HTTP requests and get back information through JSON. Added two endpoints `/api/games` and `api/games/<int:pk>`. The former gets a list view of Games and support GET and POST requests. The latter gets the detail view of a specific game using the game's `id` and supports GET, PATCH, and DELETE requests. I also added pytests to test the APIs functionality at these 2 endpoints.

@abrahmasandra worked on creating the API views as well as the testing

majorsylvie commented:
Issue Score: Excellent

Comments:
Absolutely god damn incredible work.
Your summary is fantastic, making it very very clear exactly what you implemented and what your goal is.
I particularly appreciate you detailing the relevant requests per endpoint and what they do, very readable for someone less familiar with endpoints (or at least needing their memory jogged).

I am super happy you added tests that cover multiple scenarios for your endpoint, in the future you can definitely add these in a whole nother issue and PR, I understand how circumstances led to you putting them in this PR.

Fantastic job taking all the feedback given, even the smaller not-quite-functionality-based ones about test helper functions and code comments. 

You've put in a significant amount of work, implemented very important features, helped set the tone and structure of the API app, and done it all while following the software design practices we want you to be using.

Some very small points of improvement would be adding in more formatting to your comments (such as the inline code I see you use in the closing comment), then **adding links to relevant issues and commits as needed using the `#[issue/PR number]` syntax (as well as pasting the commit SHA number, eg: 5d748fc).
Then in code, splitting thicker code like dictionaries across multiple lines, and adding more code comments (though this code does not really need more).

**Be proud of the work you've done and keep moving forward!!**

<!-- {"score": "excellent", "minor": false} -->
